kyuubi-main/src/test/scala/org/apache/kyuubi/operation/WithMiniYarnCluster.scala

```diff
@@ -0,0 +1,167 @@
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+
+import com.google.common.io.Files
+import org.apache.hadoop.yarn.conf.YarnConfiguration
+import org.apache.hadoop.yarn.server.MiniYARNCluster
+
+import org.apache.kyuubi.{KyuubiFunSuite, Utils}
+
+trait WithMiniYarnCluster extends KyuubiFunSuite {
+  // log4j configuration for the YARN containers, so that their output is collected
+  // by YARN instead of trying to overwrite unit-tests.log.
+  protected val LOG4J_CONF = """
+    |log4j.rootCategory=DEBUG, console
+    |log4j.appender.console=org.apache.log4j.ConsoleAppender
+    |log4j.appender.console.target=System.err
+    |log4j.appender.console.layout=org.apache.log4j.PatternLayout
+    |log4j.appender.console.layout.ConversionPattern=%d{yy/MM/dd HH:mm:ss} %p %c{1}: %m%n
+    |log4j.logger.org.apache.hadoop=WARN
+    |log4j.logger.org.eclipse.jetty=WARN
+    |log4j.logger.org.mortbay=WARN
+    |log4j.logger.org.sparkproject.jetty=WARN
+    """.stripMargin
+
+  private var yarnCluster: MiniYARNCluster = _
+  protected var tempDir: File = _
+  protected var hadoopConfDir: File = _
+  private var logConfDir: File = _
+
+  protected val startMiniYarnCluster: Boolean
+
+  def newYarnConfig(): YarnConfiguration = {
+    new YarnConfiguration()
+  }
+
+  def yarnClusterConfig(): Map[String, String] = {
+    yarnCluster.getConfig.iterator().asScala.map(e => (e.getKey, e.getValue)).toMap
+  }
+
+  override def beforeAll(): Unit = {
+    super.beforeAll()
+
+    if (startMiniYarnCluster) {
+      tempDir = Utils.createTempDir().toFile
+      logConfDir = new File(tempDir, "log4j")
+      logConfDir.mkdir()
+
+      val logConfFile = new File(logConfDir, "log4j.properties")
+      Files.write(LOG4J_CONF.getBytes(StandardCharsets.UTF_8), logConfFile)
+
+      // Disable the disk utilization check to avoid the test hanging when people's disks are
+      // getting full.
+      val yarnConf = newYarnConfig()
+      yarnConf.set("yarn.nodemanager.disk-health-checker.max-disk-utilization-per-disk-percentage",
+        "100.0")
+
+      // capacity-scheduler.xml is missing in hadoop-client-minicluster so this is a workaround
+      yarnConf.set("yarn.scheduler.capacity.root.queues", "default")
+      yarnConf.setInt("yarn.scheduler.capacity.root.default.capacity", 100)
+      yarnConf.setFloat("yarn.scheduler.capacity.root.default.user-limit-factor", 1)
+      yarnConf.setInt("yarn.scheduler.capacity.root.default.maximum-capacity", 100)
+      yarnConf.set("yarn.scheduler.capacity.root.default.state", "RUNNING")
+      yarnConf.set("yarn.scheduler.capacity.root.default.acl_submit_applications", "*")
+      yarnConf.set("yarn.scheduler.capacity.root.default.acl_administer_queue", "*")
+      yarnConf.setInt("yarn.scheduler.capacity.node-locality-delay", -1)
+
+      // Set RM bind host to localhost to avoid java.net.BindException
+      yarnConf.set("yarn.resourcemanager.bind-host", "localhost")
```

Comment:
try set rm bind host to localhost to avoid java.net.BindException

Reply:
set more `bind-host` to localhost.

yarn.nodemanager.bind-host
yarn.web-proxy.bind-host
yarn.router.bind-host

Refer: https://github.com/apache/hadoop/blob/26b8f678b2af98a5812c2d116e724159e37c72dd/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-common/src/main/resources/yarn-default.xml